lib/pure/osproc.nim

```diff
@@ -434,6 +434,87 @@ when not defined(useNimRtl):
 template streamAccess(p) =
   assert poParentStreams notin p.options, "API usage error: stream access not allowed when you use poParentStreams"
 
+# Wrap pipe for reading with PipeOutStream so that you can use peek* and generate runtime error
+# when setPosition/getPosition is called or write operation is performed.
+type
+  PipeOutStream[T] = ref object of T
+    buffer: Deque[char]
+    baseReadLineImpl: typeof(StreamObj.readLineImpl)
+    baseReadDataImpl: typeof(StreamObj.readDataImpl)
+
+proc posReadLine[T](s: Stream, line: var TaintedString): bool =
+  var s = PipeOutStream[T](s)
+  assert s.baseReadLineImpl != nil
+
+  let n = s.buffer.len
+  line.string.setLen(0)
+  for i in 0..<n:
+    var c = s.buffer.popFirst
+    if c == '\c':
+      c = readChar(s)
+      return true
+    elif c == '\L': return true
+    elif c == '\0':
```

Comment:
Why is a null checked for here? What if the stream contains null characters?

Reply:
This code was also copied from `readLine` proc in `streams` module.